feat(workflow): add nia-e2e workflow to setup anbox appliance and run e2e tests with adb [WD-26945]

tarek-hanna opened the pull request:
## Done

- This PR is to implement `nia-e2e.yaml` workflow and actions to setup and configure the Anbox Cloud Appliance, run Android image with adb connect on the `standard gh-hosted runner`, and then execute e2e test.

- The entrypoint for these actions, is the `nia-pr.yaml` workflow triggered on PR which calls 2 workflows:

1. `nia-ci.yaml` :

- runs lint, test and build nia apk

2. `nia-e2e.yaml`:  

- installs LXD and Anbox Cloud Appliance using `canonical/anbox-cloud-github-action`.
- creates an android jammy image and connects with adb.
- runs the e2e tests `./gradlew  connectedDemoDebugAndroidTest`.
- upload test report as e2e artifacts.

## QA
Perform the following QA steps:
- Create or update this PR.
- All checks should re-run and complete successfully
- Artifacts should include nia-e2e test reports.

## JIRA / Launchpad bug
Fixes # 
1- [Anbox Cloud Demo - Add nia-e2e steps to setup anbox appliance and create android instance](https://warthogs.atlassian.net/browse/WD-25541)
2- [Anbox Cloud Demo - Add nia-e2e step to run e2e tests](https://warthogs.atlassian.net/browse/WD-26945) 


## Documentation
- README.md doc is updated to explain how the new workflow and actions work.


## Screenshots
**E2E Workflow Steps**
<img width="1920" height="750" alt="Screenshot from 2025-09-23 21-28-33" src="https://github.com/user-attachments/assets/d9d78805-9880-48e0-b559-3667ad012ad1" />

**E2E Connected Demo Debug Android Test Report - Success Scenario**
<img width="1186" height="406" alt="Screenshot from 2025-09-21 11-03-55" src="https://github.com/user-attachments/assets/11c63cf1-74bc-4899-b80b-3c0c833c7611" />

**E2E Connected Demo Debug Android Test Report - Failure Scenario**
<img width="1186" height="406" alt="Screenshot from 2025-09-21 11-16-42" src="https://github.com/user-attachments/assets/170efaec-db0e-4117-9362-139d9f74611b" />

ajanon commented on .github/actions/setup-adb/action.yaml:
```diff
@@ -0,0 +1,46 @@
+  inputs:
+    android-version:
+      default: 15
+      description: Set jammy:android version to use
```
Is this description up-to-date? From the name of the input, I would expect it to be only the android version.

tarek-hanna replied:
@ajanon  yes it is only the version number as in " jammy:android${ANDROID_VERSION}" .. Do you have in mind a different format ?

ajanon replied:
Ah, yes, I just feels liks it's not necessary to explain the format here.
I think something like the following could be sufficient:

```suggestion
      description: Set android version to use
```

Wdyt?

Also, we may want to be able to add the Ubuntu version as an additional input. For now, we only support jammy, but this may (will) change in the future.

tarek-hanna replied:
That's a good idea, would also give flexibility to run a matrix with multiple versions  & images

tarek-hanna replied:
Done

ajanon reviewed (commented):
Just one nitpick

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+    - name: Setup Anbox Cloud Appliance
+      uses: ./.github/actions/setup-anbox
+      with:
+        ubuntu-pro-token: ${{ secrets.UBUNTU_PRO_TOKEN }}
```
This wont fly. As this is a public repo we wont be able to run the workflow for PRs from people without write access. This is why we have the cache workflow in https://github.com/canonical/anbox-cloud-github-action/blob/main/.github/workflows/cache-images.yaml Any reason you didn't use this here?

tarek-hanna replied:
I checked this one and it also requires UBUNTU_PRO_TOKEN .. but i found this [image-cache used in streaming sdk](https://github.com/canonical/anbox-streaming-sdk/blob/main/.github/workflows/cache-images.yaml).. i will try it and resubmit the PR

morphis replied:
> I checked this one and it also requires UBUNTU_PRO_TOKEN .. but i found this [image-cache used in streaming sdk](https://github.com/canonical/anbox-streaming-sdk/blob/main/.github/workflows/cache-images.yaml).. i will try it and resubmit the PR

Yes it does require the token but it runs on schedule, not on PR. That's the difference and when running on schedule we can safely use secrets :-)

tarek-hanna replied:
Done

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+      uses: ./.github/actions/setup-android
+    
+    - name: Setup Anbox Cloud Appliance
+      uses: ./.github/actions/setup-anbox
```
As discussed we should keep this as bare and as close to the official steps we document for the action and in our documentation as possible. These are nice additions but not necessarily something which will fit for people looking at the example. Even if that makes the workflow longer that is what we're looking for to show as an example. If there things we can do in the official action to reduce boilerplate, lets do that instead.

tarek-hanna replied:
ok i will change that in nia-e2e.yaml and nia-ci.yaml as well.

tarek-hanna replied:
Android SDK JDK Installation steps now removed.. it was only needed in my first attempt on locally hosted runner, but now not needed anymore.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+      uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+    
+    - name: Setup Android Dependencies
+      uses: ./.github/actions/setup-android
```
I am not sure how this slipped through but the authorship for the uses https://github.com/android-actions is not clear to me and we should refrain for the example from using any 3rd party actions other than what comes from Github (https://github.com/actions). As also discussed we can assume and require GitHub runner compatible images which already should come with everything we need. Was there a specific reason for which you had to add this?

tarek-hanna replied:
removed

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+        ubuntu-pro-token: ${{ secrets.UBUNTU_PRO_TOKEN }}
+
+    - name: Create Android Instance and Connect ADB
+      uses: ./.github/actions/setup-adb
```
See above.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+  nia-e2e:
+    name: Setup Anbox and Run E2E Tests
+    runs-on: ubuntu-latest
+    if: contains(github.event.pull_request.labels.*.name, 'run-nia-e2e-tests')
```
We should run this for every PR changing something for nowinandroid and not only when the label is set.

tarek-hanna replied:
removed. triggers in PR and needs lint, unit test and build job to succeed first

morphis commented on .github/actions/setup-anbox/action.yaml:
```diff
@@ -0,0 +1,30 @@
+      shell: bash
+      run: |
+        amc config set container.security_updates false
+        amc node set lxd0 cpu-allocation-rate 10
```
What's the reason for this?

tarek-hanna replied:
i added it to fix an issue on my local runner used in ams TICS as well, but i removed it

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,49 @@
+    timeout-minutes: 60
+    defaults:
+      run:
+        working-directory: ./nowinandroid
```
If you set a default working-directory I don't think the `uses: ./.github/...` below will work.

tarek-hanna replied:
it is working, but looks like we will not need any of the actions anyway, so it shouldn't be a problem

ajanon reviewed (approved):
LGTM

morphis commented on .github/workflows/cache-images.yaml:
```diff
@@ -0,0 +1,49 @@
+name: Cache Anbox Cloud images
+
+on:
+  pull_request:
```
This cannot run on a pull request as there is no secret access in that case.

morphis commented on .github/workflows/cache-images.yaml:
```diff
@@ -0,0 +1,49 @@
+
+on:
+  pull_request:
+  workflow_dispatch:
```
If we run this on a daily basis any manual invocation will be quickly overwritten unless we add the Android version to the cache artifact.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,78 @@
+        test "$(adb shell getprop ro.product.model)" = Anbox
+
+    - name: Run e2e tests
+      run: ./gradlew :app:installDemoDebug
```
This only installs the APK but doesn't run any tests, we need `connectedDemoDebugAndroidTest` for that see https://github.com/android/nowinandroid?tab=readme-ov-file#testing

tarek-hanna replied:
Done

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,78 @@
+  workflow_call:
+    inputs:
+      android-version:
+        default: "13"
```
Why not use 15 by default to show the latest we have available?

tarek-hanna replied:
"15" Done

morphis commented on .github/workflows/cache-images.yaml:
```diff
@@ -0,0 +1,49 @@
+    name: Cache Anbox Cloud images
+    runs-on: ubuntu-latest
+    env:
+        ANDROID_VERSION: ${{ inputs.android-version || '13' }}
```
Lets use 15, wich is the latest we have available.

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,78 @@
+        type: string
+      use-vm:
+        default: "false"
+        description: Run Android in a VM instead of a system container
```
This not precisely what this is. We run Android in a system container in a VM in this case.

tarek-hanna replied:
removed. No inputs needed

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,73 @@
+
+
+
+    
```
Can you drop superfluous new lines?

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,73 @@
+
+    - name: Import cached images
+      run: |
+        set -x
```
Any reason we need this?

tarek-hanna replied:
for logging. will remove :+1:

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,73 @@
+    - name: Create android instance
+      id: create-instance
+      run: |
+        set -x
```
Same as above

tarek-hanna replied:
was already there from https://github.com/canonical/anbox-cloud-github-action/blob/main/.github/workflows/pr.yaml. 

removed :+1:

morphis commented on .github/workflows/nia-e2e.yaml:
```diff
@@ -0,0 +1,66 @@
+      uses: actions/cache/restore@5a3ec84eff668545956fd18022155c47e93e2684 # v4
+      with:
+        path: images
+        key: anbox-demos-images-amd64
```
The name is just `anbox-images-amd64`, that's also why the workflow fails:

<img width="1896" height="1083" alt="Image" src="https://github.com/user-attachments/assets/ee6160ce-e4ff-4749-87f6-45578f100406" />

<img width="1903" height="263" alt="Image" src="https://github.com/user-attachments/assets/2f6ca185-eb9a-4677-ae3c-36e3490a1eaf" />

tarek-hanna replied:
it was `anbox-images-amd64` and it kept failing so i changed it to `anbox-demos-images-amd64` that you see now. It wont work unless the 'cache-image.yaml` workflow is merged on main and dispatched.

tarek-hanna replied:
i reverted back the key but it will fail again

tarek-hanna committed:
feat: implement nia-e2e.yaml workflow to setup  anbox cloud appliance and connect with ADB

This commit is to implement nia-e2e.yaml workflow to setup and configure
anbox cloud appliance, and run android image with ADB on the gh-hosted
runner.

nia-e2e is triggered from the pr workflow after nia-ci successfully
completes.

nia-e2e uses anbox-cloud-github-actions to install and setup lxd and
anbox appliance, and then restores a cached android: jammy image to run.

e2e will run connectedDemoDebugAndroidTest to check for adb connection
and then upload the test report as artifacts

README.md doc is also updated to demonstrate the workflow and how it
works.

tarek-hanna committed:
fix: set working directory on step level

morphis reviewed (approved):
Great to see this finally working! Great work @tarek-hanna!
